staging/src/k8s.io/component-base/featuregate/testing/feature_gate.go

```diff
@@ -18,18 +18,54 @@ package testing
 
 import (
 	"fmt"
-	"testing"
+	"strings"
+	"sync"
 
 	"k8s.io/component-base/featuregate"
 )
 
-// SetFeatureGateDuringTest sets the specified gate to the specified value, and returns a function that restores the original value.
-// Failures to set or restore cause the test to fail.
+var (
+	overwrittenFeatureFlags map[featuregate.Feature]string
+	lock                    sync.Mutex
+)
+
+func init() {
+	overwrittenFeatureFlags = map[featuregate.Feature]string{}
+}
+
+// SetFeatureGateDuringTest sets the specified gate to the specified value for duration of the test.
+// Fails when it detects second call to the same flag or is unable to set or restore feature flag.
+// Returns empty cleanup function to maintain the old function signature that uses defer.
+// TODO: Remove defer from calls to SetFeatureGateDuringTest and update hack/verify-test-featuregates.sh when we can do large scale code change.
+//
+// WARNING: Can leak set variable when called in test calling t.Parallel(), however second attempt to set the same feature flag will cause fatal.
 //
 // Example use:
 //
 // defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, true)()
-func SetFeatureGateDuringTest(tb testing.TB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+func SetFeatureGateDuringTest(tb testingTB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
```

Comment:
I don't think the override detection is correct... it looks like serial calls to SetFeatureGateDuringTest in a test would trip it:

```
SetFeatureGateDuringTest(..., true)
...
SetFeatureGateDuringTest(..., false)
```

can we simplify this to make sure we don't interleave calls to SetFeatureGateDuringTest from different / non-descendant tests based on the [name of the test](https://pkg.go.dev/testing#F.Name) this was called from until cleanup is done?

var (
  testStack []string
  testStackLock sync.Mutex
)

func SetFeatureGateDuringTest(tb testingTB...) {
  func (){
    testStackLock.Lock()
    defer testStackLock.Unlock()
    if len(testStack) > 0 {
      lastCalledFrom = testStack[len(testStack)-1]
      if lastCalledFrom != tb.Name() && strings.HasPrefix(tb.Name(), lastCalledFrom + "/") {
        tb.Errorf("...")
        return
      }
    }
    testStack = append(testStack, tb.Name())
  }()


  return func() {
    testStackLock.Lock()
    defer testStackLock.Unlock()
    // make sure we're the last item on the stack
    if len(lastStack) == 0 || testStack[len(testStack)-1] != tb.Name() {
      tb.Errorf("...")
    }
    testStack = testStack[0:len(testStack)-2]

    ...
  }
...

Reply:
Serial calls will not trip it. Please check tests. 

I would need to double check, but your implementation doesn't seem to work too different than mine, the difference is that you have an explicit stack, while in my stack is implicitly stored  in function closures. See that I pass the `originalValue`, `previousTestName` and `alreadyOverwritten` from the function call to the cleanup function. I think it should work as a stack, haven't compared the details of your implementation.

Reply:
ah, I missed the closure-based stack, sorry, I was looking at an older version of the PR

Reply:
I *think* the closure solution is right, though I find it harder to follow than an actual stack

Reply:
Maybe it's easier as your code assumes single feature flag :P

Reply:
Also, my code needs cleanup

Reply:
Done, ptal